tawazi/ops.py

```diff
@@ -53,40 +64,107 @@ def to_dag(
     *,
     max_concurrency: int = 1,
     behavior: ErrorStrategy = ErrorStrategy.strict,
-) -> Callable[..., Any]:
+) -> DAG:
     """
-    Transform the declared ops into a DAG that can be executed by tawazi.
+    Transform the declared ops into a DAG that can be executed by tawazi's scheduler.
     The same DAG can be executed multiple times.
     Note: to_dag is thread safe because it uses an internal lock.
         If you need to construct lots of DAGs in multiple threads,
-        it is best to construct your dag once and then consume it as much as you like in multiple threads.
+        it is best to construct your dag once and then use it as much as you like.
     Please check the example in the README for a guide to the usage.
     Args:
         declare_dag_function: a function that contains the execution of the DAG.
-        if the functions are decorated with @op decorator they can be executed in parallel.
-        Otherwise, they will be executed immediately. Currently mixing the two behaviors isn't supported.
-        max_concurrency: the maximum number of concurrent threads to execute in parallel
-        behavior: the behavior of the dag when an error occurs during the execution of a function (ExecNode)
+        Currently Only @op decorated functions can be used inside the decorated function (i.e. declare_dag_function).
+        However, you can use some simple python code to generate constants.
+        max_concurrency: the maximum number of concurrent threads to execute in parallel.
+        behavior: the behavior of the DAG when an error occurs during the execution of a function (ExecNode).
     Returns: a DAG instance
-
     """
 
-    def intermediate_wrapper(_func: Callable[..., Any]) -> Callable[..., DAG]:
-        def wrapped(*args: Any, **kwargs: Any) -> DAG:
-            # TODO: modify this horrible pattern
-            with exec_nodes_lock:
-                node.exec_nodes = []
-                _func(*args, **kwargs)
+    # wrapper used to support parametrized and non parametrized decorators
+    def intermediate_wrapper(_func: Callable[..., Any]) -> DAG:
+
+        # 0. Protect against multiple threads declaring many DAGs at the same time
+        with exec_nodes_lock:
+
+            # 1. node.exec_nodes contains all the ExecNodes that concern the DAG being built at the moment.
+            #      make sure it is empty
+            node.exec_nodes = []
+            try:
```

Comment:
as I said I am not a huge fan of this pattern but I don't have any better idea at the moment

Reply:
are you talking about the try: ... finally: ... ?

Reply:
yep this one